node/src/components/diagnostics_port.rs

```diff
@@ -0,0 +1,247 @@
+#[derive(Clone, DataSize, Debug, Deserialize)]
+pub(crate) struct Config {
+    /// Whether or not the diagnostics port is enabled.
+    enabled: bool,
+    /// Path to listen on.
+    socket_path: PathBuf,
+    /// `umask` to apply before creating the socket.
+    socket_umask: u16,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            enabled: false,
+            socket_path: "debug.socket".into(),
+            socket_umask: 0o077,
+        }
+    }
+}
+
+impl DiagnosticsPort {
+    /// Creates a new diagnostics port component.
+    pub(crate) fn new<REv>(
+        cfg: &WithDir<Config>,
+        event_queue: EventQueueHandle<REv>,
+    ) -> Result<(Self, Effects<Event>), Error>
+    where
+        REv: From<DumpConsensusStateRequest> + From<ControlAnnouncement> + Send,
+    {
+        let config = cfg.value();
+        let (shutdown_sender, shutdown_receiver) = watch::channel(());
+
+        if !config.enabled {
+            // If not enabled, do not launch a background task, simply exit immediately.
+            //
+            // Having a shutdown sender around still is harmless.
+            debug!("diagnostics port disabled");
+            return Ok((DiagnosticsPort { shutdown_sender }, Effects::new()));
+        }
+
+        let socket_path = cfg.with_dir(config.socket_path.clone());
+        let listener = setup_listener(
+            &socket_path,
+            // Mac OS X / Linux use different types for the mask, so we need to call .into() here.
+            #[allow(clippy::useless_conversion)]
+            config.socket_umask.into(),
+        )?;
+        let server = tasks::server(
+            EffectBuilder::new(event_queue),
+            socket_path,
+            listener,
+            shutdown_receiver,
+        );
+
+        Ok((DiagnosticsPort { shutdown_sender }, server.ignore()))
+    }
+}
+
+/// Sets up a UNIX socket listener at the given path.
+///
+/// If the socket already exists, an attempt to delete it is made. Errors during deletion are
+/// ignored, but may cause the subsequent socket opening to fail.
+fn setup_listener<P: AsRef<Path>>(path: P, socket_umask: umask::Mode) -> io::Result<UnixListener> {
+    let socket_path = path.as_ref();
+
+    // This would be racy, but no one is racing us for the socket, so we'll just do a naive
+    // check-then-delete :).
+    if socket_path.exists() {
+        debug!(socket_path=%socket_path.display(), "found stale socket file, trying to remove");
+        match fs::remove_file(&socket_path) {
+            Ok(_) => {
+                debug!("stale socket file removed");
+            }
+            Err(err) => {
+                // This happens if a background tasks races us for the removal, as it usually
+                // means the file is already gone. We can ignore this, but make note of it in
+                // the log.
+                debug!(%err, "could not remove stale socket file, assuming race with background task");
```

Comment:
Should this be `warn` at least?

Reply:
It could be external to the node as well. I'll upgrade the warning (and fix the typo in the comment as well :))